execution_workflows/Aptardi/bin/make_identification_bed.py

```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python
+
+import sys
+
+input_gtf_path=sys.argv[1]
+input_gtf=open(input_gtf_path,"r")
+output_bed_path=sys.argv[2]
+if output_bed_path == 'not_specified':
+ output_bed=open(input_gtf_path.split(".")[0]+"_identification.bed","w")
+else:
+ output_bed=open(output_bed_path,"w")
+
+for ln in input_gtf:
+ if not ln.startswith("#"):
+  ln=ln.strip().split("\t")
+  if ln[1] == "aptardi": #get the Aptardi identified transcripts only
+   attrList=ln[-1].split(";")
+   attrDict={}
+   for k in attrList:
+    p=k.strip().split(" ")
+    if len(p) == 2:
+     attrDict[p[0]]=p[1].strip('\"')
+   name=attrDict["transcript_id"]
+   #chrom,chromStart,chromEnd,name,score,strand
+   output_bed.write("\t".join([ln[0],ln[3],ln[4],name,".",ln[6]])+"\n")
+output_bed.close()
```

Comment:
The `Aptardi` output consists of both the `Aptardi`-found entries and the entries from the original `gtf` file, and if I include entries with `transcripts` in the third field, then I will be including entries from the original `gtf` file, not the `Aptardi`-identified sites.
I think the conversion of the `Aptardi` output to poly(A) sites is not that straightforward, given the `Aptardi` entries are several hundred nucleotides long, which will require a database for annotating them.

Reply:
Could you also select 'transcript' rows that have 'aptardi' in the 2nd column ('source' column I believe?)? My impression is that these are the Aptardi predicted transcripts, and the 3'ends of each predicted transcript correspond to the predicted poly(A) site. So essentially all we need to get a polyA site is to take the [3'most coordinate to represent the cleavage site](https://github.com/iRNA-COSI/APAeval/issues/159).

As a rough pass you could do something like this using [PyRanges](), which (I think) is quite straightforward:

```
import pyranges as pr
import pandas as pd

gtf = pr.read_gtf("path/to/file.gtf")

# Source is 2nd col in GTF, Feature is 3rd col
gtf = gtf.subset(lambda df: (df["Source"] == "aptardi") & (df["Feature"] == "transcript"))

# Strand aware, reports 3'most coordinate for each interval
# PyRanges coordinates follow BED conventions
three_ends = gtf.three_end()

# Some reformatting to get a BED6 file

three_ends = three_ends[["Strand"]]

# Dummy score column
three_ends.Score = "."

# Making up a Name column, can be whatever you like though
# <chr>:<start>:<end>
three_ends = three_ends.assign("Name",
lambda df: df[["Chromosome","Start","End"]].apply(lambda row: ":".join(row.values.astype(str)), axis=1))

three_ends.to_bed("output.bed")

```
I've just picked out PyRanges because I'm familiar with the package. I haven't tested that code but it should be functional, mainly was a means to show you what I had in mind? Does the workflow make sense?

Reply:
Thanks for the code, @SamBryce-Smith! I have updated this using your code and [the Docker image](https://hub.docker.com/r/apaeval/aptardi/tags). Here is the run:
![Screenshot from 2021-12-13 15-46-54](https://user-images.githubusercontent.com/41866052/145780326-7ee5c4c7-76e3-4adf-8db6-d3363caaad81.png)

Reply:
Amazed it worked first time :O